graspy/inference/nonpar.py

```diff
@@ -0,0 +1,144 @@
+    ):
+
+        if type(n_bootstraps) is not int:
+            raise TypeError()
+
+        if bandwidth is not None and type(bandwidth) is not float:
+            raise TypeError()
+
+        if n_bootstraps < 1:
+            msg = "{} is invalid number of bootstraps, must be greater than 1"
+            raise ValueError(msg.format(n_bootstraps))
+
+        super().__init__(embedding=embedding, n_components=n_components)
+        self.n_bootstraps = n_bootstraps
+        self.bandwidth = bandwidth
+        self.embedding = embedding
+
+    def _gaussian_covariance(self, X, Y):
+        diffs = np.expand_dims(X, 1) - np.expand_dims(Y, 0)
+        if self.bandwidth == None:
+            self.bandwidth = 0.5
+        return np.exp(-0.5 * np.sum(diffs ** 2, axis=2) / self.bandwidth ** 2)
+
+    def _statistic(self, X, Y):
+        N, _ = X.shape
+        M, _ = Y.shape
+        x_stat = np.sum(self._gaussian_covariance(X, X) - np.eye(N)) / (N * (N - 1))
+        y_stat = np.sum(self._gaussian_covariance(Y, Y) - np.eye(M)) / (M * (M - 1))
+        xy_stat = np.sum(self._gaussian_covariance(X, Y)) / (N * M)
+        return x_stat - 2 * xy_stat + y_stat
+
+    def _ase(self, A, max_d):
+        ase = AdjacencySpectralEmbed(n_components=max_d, algorithm="randomized")
+        X_hat = ase.fit_transform(A)
+        return X_hat
+
+    def _lse(self, A, max_d):
+        lse = LaplacianSpectralEmbed(n_components=max_d)
+        X_hat = lse.fit_transform(A)
+        return X_hat
+
+    def _median_heuristic(self, X1, X2):
+        X1 = np.array(X1)
+        X2 = np.array(X2)
+        X1_medians = np.median(X1, axis=0)
+        X2_medians = np.median(X2, axis=0)
+        val = np.multiply(X1_medians, X2_medians)
+        t = (val > 0) * 2 - 1
+        X1 = np.multiply(t.reshape(-1, 1).T, X1)
+        return X1, X2
+
+    def _bootstrap(self, X, Y, M=200):
+        N, _ = X.shape
+        M2, _ = Y.shape
+        Z = np.concatenate((X, Y))
+        statistics = np.zeros(M)
+        for i in range(M):
+            bs_Z = Z[
+                np.random.choice(np.arange(0, N + M2), size=int(N + M2), replace=False)
+            ]
+            bs_X2 = bs_Z[:N, :]
+            bs_Y2 = bs_Z[N:, :]
+            statistics[i] = self._statistic(bs_X2, bs_Y2)
+        return statistics
+
+    def fit(self, A1, A2):
+        """
+        Fits the test to the two input graphs
+
+        Parameters
+        ----------
+        A1, A2 : nx.Graph, nx.DiGraph, nx.MultiDiGraph, nx.MultiGraph, np.ndarray
+            The two graphs to run a hypothesis test on.
+
+        Returns
+        -------
+        p : float
+            The p value corresponding to the specified hypothesis test
+        """
+        A1 = symmetrize(import_graph(A1))
+        A2 = symmetrize(import_graph(A2))
+        # if type(A1) != 'numpy.ndarray' or type(A2) != 'numpy.ndarray':
+        #     return TypeError()
+        A1_d = select_dimension(A1)[0][-1]
```

Comment:
Wrap this around if self.n_components is None. This doesn't allow users to set the embedding dimensions.